### PR TITLE
Add a Markdown parser to convert Discord formatting to Minecraft

### PR DESCRIPTION
**Pull request by EbonJaeger**
There is one known caveat so far: Text consisting of italics with an inner bold element breaks. For example:
`*Italic **Bold***`

Some more tests and better documentation also needs to be done, and probably a cleanup pass.

### CODE REVIEW

**Comment by EbonJaeger** [comment]
Lots more tests have been added. With the last commit, all of the edge cases I could think of now work correctly, verified by the code tests and in-game testing.

File: src/discord.rs [hunk 1]
```diff
@@ -303,15 +305,28 @@ fn truncate_lines<'a>(lines: Split<'a, &'a str>) -> Vec<String> {
...
             };
         }
     }
 
     truncated
 }
 
+/// Parses lines of text for Discord markdown so that it can be
+/// translated into Minecraft chat formatting.
+fn parse_markdown(lines: Split<'_, char>) -> Vec<String> {
```
Comment (EbonJaeger) [inline src/discord.rs]: Do we really need to have a function for this?

File: src/discord.rs [hunk 2]
```diff
@@ -483,8 +498,11 @@ fn split_webhook_url(url: &str) -> Option<(u64, &str)> {
     }
 
     let captures = match WEBHOOK_REGEX.captures(&url) {
-        Some(captures) => captures,
-        None => return None,
+        Ok(result) => match result {
+            Some(captures) => captures,
+            None => return None,
+        },
+        Err(_) => return None,
```
Comment (EbonJaeger) [inline src/discord.rs]: This error should be logged, even if it's not actually useful.

File: src/markdown/block/blockquote.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
+use super::Block;
+use super::Block::Blockquote;
+use crate::markdown::block::parse_blocks;
+
+pub fn parse_blockquote(lines: &[&str]) -> Option<(Block, usize)> {
```
Comment (EbonJaeger) [inline src/markdown/block/blockquote.rs]: This shouldn't ever see multiple lines. With that assumption out of the way, the function logic should be vastly simplified.

File: src/markdown/block/mod.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+use super::Span::Text;
+
+mod blockquote;
+
+use self::blockquote::parse_blockquote;
+
+pub fn parse_blocks(content: &str) -> Vec<Block> {
+    let mut blocks = vec![];
+    let mut t = vec![];
+    let lines: Vec<&str> = content.lines().collect();
```
Comment (EbonJaeger) [inline src/markdown/block/mod.rs]: We can safely assume that we're only passing one single line to this function.

File: src/markdown/mod.rs [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+fn format_blockquote(elements: &[Block]) -> String {
+    format!("> {}", to_minecraft_format(elements))
+}
+
+fn format_paragraph(elements: &[Span]) -> String {
+    format_spans(elements, &mut vec![])
+}
+
+fn format_spans(elements: &[Span], mut open_tags: &mut Vec<String>) -> String {
```
Comment (EbonJaeger) [inline src/markdown/mod.rs:96]: Add some documentation about what the whole `open_tags` thing is about and how it works. Can we also possibly make those strings constants?

File: src/markdown/span/emphasis.rs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        static ref EMPHASIS: Regex = Regex::new(r"^\b_((?:__|\\[\s\S]|[^\\_])+?)_\b|^\*(?=\S)((?:\*\*|\\[\s\S]|\s+(?:\\[\s\S]|[^\s\*\\]|\*\*)|[^\s\*\\])+?)\*(?!\*)").unwrap();
+    }
+
+    match EMPHASIS.is_match(text) {
+        Ok(matches) => {
+            if matches {
+                let captures = EMPHASIS
+                    .captures(text)
+                    .expect("error running regex")
+                    .expect("no match found");
```
Comment (EbonJaeger) [inline src/markdown/span/emphasis.rs]: Maybe instead of doing this, and checking if we have matches in the above `if` statement, we could use matching instead. That would make sure that we're not running the regex twice for some reason (once to check for matches, and again to get the captures; I'm not sure how this regex crate works in this respect).

File: src/markdown/span/strikethrough.rs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+use crate::markdown::Span::Strikethrough;
+use fancy_regex::Regex;
+
+/// Parses any strikethrough markdown tags in the given text.
+pub fn parse_strikethrough(text: &str) -> Option<(Span, usize)> {
+    lazy_static! {
+        static ref STRIKETHROUGH: Regex = Regex::new(r"^~~(?P<text>.+?)~~").unwrap();
+    }
+
+    match STRIKETHROUGH.is_match(text) {
```
Comment (EbonJaeger) [inline src/markdown/span/strikethrough.rs]: See comment in `emphasis.rs`.

File: src/markdown/span/strong.rs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+use crate::markdown::Span::Strong;
+use fancy_regex::Regex;
+
+/// Parses any strong (bold) markdown tags in the given text.
+pub fn parse_strong(text: &str) -> Option<(Span, usize)> {
+    lazy_static! {
+        static ref STRONG: Regex = Regex::new(r"^\*\*(?P<text>.+?)\*\*(?!\*)").unwrap();
+    }
+
+    match STRONG.is_match(text) {
```
Comment (EbonJaeger) [inline src/markdown/span/strong.rs]: See comment in `emphasis.rs`.

File: src/markdown/span/underline.rs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+use crate::markdown::Span::Underline;
+use fancy_regex::Regex;
+
+/// Parses any underline markdown tags in the given text.
+pub fn parse_underline(text: &str) -> Option<(Span, usize)> {
+    lazy_static! {
+        static ref UNDERLINE: Regex = Regex::new(r"^__(?P<text>.+?)__(?!_)").unwrap();
+    }
+
+    match UNDERLINE.is_match(text) {
```
Comment (EbonJaeger) [inline src/markdown/span/underline.rs]: See comment in `emphasis.rs`.

**Review by EbonJaeger — COMMENTED** [review]
Just some (hopefully) minor cleanup issues to take care of. Also remember to run `cargo fmt` and `cargo clippy` before merging.

File: src/discord.rs (hunk 2 above)
Comment (EbonJaeger) [reply src/discord.rs]: Eh. We do the same thing right below it anyways.
